src/index.rs

```diff
@@ -85,6 +85,76 @@ pub struct IndexEntry {
     pub path: Vec<u8>,
 }
 
+// We cannot return raw::git_index_entry instances directly, as they rely on
+// a CString which is owned by the function. To make the pointer to the CString
+// valid during usage of raw::git_index_entry, we supply the index entry in a
+// callback where pointers to the CString are valid.
+fn try_raw_entries(
```

Comment:
I think this function would be easier to use as:

```rust
impl IndexEntry {
    fn to_raw(&self, storage: &mut Vec<CString>) -> raw::git_index_entry {
        // ...
    }
}
```

which would avoid the vector, allocations internally, and callback style

Reply:
In my approach the `Vec`s are necessary as I need multiple entries available in the callback, therefore `try_raw_entries` can handle multiple entries.

With your approach I think something like this would work:

```
impl IndexEntry {
    fn to_raw(&self, cpath: &CString) -> raw::git_index_entry {
        // ...
    }
}
```

Although, I think this would still not ensure that `cpath` lives longer than the returned value. The usage would look like this:

```
let cpath = CString::new(&self.path[..])?;
let raw = self.to_raw(&cpath);
```

My last commit replaces the `Vec`s with arrays but still contains the callback. If it is safe to have a pointer pointing to the `cpath` parameter in the return value, I would convert the implementation to that approach. But I'm unsure whether `cpath` is actually always freed after the return value as it has no lifetime attached to it. I am not that familiar with the internals of memory management in Rust.

Reply:
I implemented your callback-less approach and checked if the tests are passing. Looks good so far, but I'm still not sure whether this API is safe: the return value references data of a parameter but does not carry the lifetime of the parameter to indicate that `c_path` must live longer than the returned value. Maybe this is not necessary as the `c_path` binding is always created before the return value in assigned to a variable, I don't know.